Allow getting logs directly from deployment, job and statefulset

soltysh opened the pull request:
**Special notes for your reviewer**:
@smarterclayton you asked for it in OpenShift


```release-note
kubectl logs allows getting logs directly from deployment, job and statefulset
```

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/40927)
<!-- Reviewable:end -->

soltysh commented:
@kubernetes/sig-apps-pr-reviews

pwittrock commented:
Do tests exist for these?

soltysh commented:
@pwittrock I haven't seen any, I'll add one.

pwittrock commented:
That would be wonderful.  Thank you.

krmayankk commented on pkg/kubectl/cmd/util/factory_object_mapping.go:
```diff
@@ -260,6 +260,26 @@ func (f *ring1Factory) LogsForObject(object, options runtime.Object) (*restclien
+			return nil, fmt.Errorf("invalid label selector: %v", err)
+		}
+		sortBy := func(pods []*v1.Pod) sort.Interface { return controller.ByLogging(pods) }
+		pod, numPods, err := GetFirstPod(clientset.Core(), t.Namespace, selector, 20*time.Second, sortBy)
```
Will this only work for the first pod ?

smarterclayton replied:
Yes, consistent with the others.  See https://github.com/kubernetes/kubernetes/issues/19873 for proposed extension.

mfojtik replied:
shouldn't we check if replicas > 0 ? AFAIK this function will wait for 20s for a Pod to appear, but I think we should return faster

smarterclayton replied:
If it's consistent with the others, why would we change the behavior here?

kargakis commented:
@soltysh I think the help message of the command needs an update

mfojtik commented on pkg/kubectl/cmd/util/factory_object_mapping.go:
```diff
@@ -260,6 +260,26 @@ func (f *ring1Factory) LogsForObject(object, options runtime.Object) (*restclien
+			return nil, err
+		}
+		if numPods > 1 {
+			fmt.Fprintf(os.Stderr, "Found %v pods, using pod/%v\n", numPods, pod.Name)
```
why is this error?

soltysh replied:
Consistently with others.

soltysh commented:
I've added more objects per @smarterclayton request, that is StatefulSet, Job and CronJob. I've also added them to `AttachablePodForObject`. I still need to figure out the test for this.

kargakis commented on pkg/kubectl/cmd/util/factory_object_mapping.go:
```diff
@@ -330,6 +390,16 @@ func (f *ring1Factory) AttachablePodForObject(object runtime.Object) (*api.Pod,
 		pod, _, err := GetFirstPod(clientset.Core(), t.Namespace, selector, 1*time.Minute, sortBy)
 		return pod, err
+
+	case *extensions.ReplicaSet:
```
There is an open PR for attach, mind leaving that aside?
https://github.com/kubernetes/kubernetes/pull/40365

soltysh replied:
Sure thing.

soltysh committed:
Switch ClientAccessFactory to use clientset.Interface rather than actual type

soltysh committed:
Allow getting logs directly from deployment, job and statefulset

soltysh commented:
@kargakis I've removed the attach part
@pwittrock added tests, although now that we have a placeholder for more tests probably `AttachablePodForObject` could deserve its own tests as well :) (as a followup)
@deads2k and @fabianofranz  you might be interested in the first commit, for the `ClientAccessFactory` to be usable for tests I had to change method signatures to use `internalclientset.Interface` instead of actual type, can one of you approve that change?
@smarterclayton ready for final review

kargakis commented on pkg/kubectl/cmd/logs.go:
```diff
@@ -54,13 +54,19 @@ var (
+		kubectl logs job/hello
+
+		# Return snapshot logs from from container hello-1 of a job named hello
+		kubectl logs job/hello -c hello-1`)
```
Provide an example with a different resource since you already have one with jobs above

kargakis commented:
The relase note needs an update

kargakis commented:
Apart from a couple of comments and the clientset interface update, lgtm

soltysh commented:
@kargakis nits fixed, release note updated

soltysh commented:
@k8s-bot kubemark e2e test this

shiywang commented:
hi @soltysh @kargakis just have a simple question here, since `logs` can support multiple type, does it support those formats below ?
``` console
kubectl logs rc rc-test 
kubectl logs rc rc-test containerName
```

shiywang commented on pkg/kubectl/cmd/logs.go:
```diff
@@ -94,9 +100,7 @@ func NewCmdLogs(f cmdutil.Factory, out io.Writer) *cobra.Command {
-			if err := o.Validate(); err != nil {
-				cmdutil.CheckErr(cmdutil.UsageError(cmd, err.Error()))
-			}
+			cmdutil.CheckErr(o.Validate())
```
why remove this ? if a user provided an invalid logs options, shouldn't we prompt this ?
```console
unexpected logs options object 
See 'logs -h' for help and examples.
```

soltysh replied:
All user input is verified in `Complete`, at this point we're validating the correctness of internal structures (here `LogsOptions`) before running the actual command. User has no real input on these values, thus this doesn't provide any value for him. See other commands for similarity.

shiywang replied:
ok, I see, thanks

soltysh commented:
> hi @soltysh @kargakis just have a simple question here, since logs can support multiple type, does it support those formats below ?
kubectl logs rc rc-test 
kubectl logs rc rc-test containerName

No, for these resources (jobs, statefulset, etc) you need to pass resource/name explicitly, to make it work.

soltysh commented:
Rebased.

kargakis commented:
Initially we couldn't support RESOURCE NAME syntax because of backwards
compatibility (second arg used to be the container name for the pod)

On Mon, Feb 20, 2017 at 12:50 PM, Maciej Szulik <notifications@github.com>
wrote:

> Rebased.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/40927#issuecomment-281059793>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADuFfyHqXgHB80pFm9xBACFAzT6-aVIfks5reX4LgaJpZM4L2f4g>
> .
>

soltysh commented:
@k8s-bot cvm gke e2e test this

soltysh commented:
@k8s-bot kubemark e2e test this

soltysh commented:
@pwittrock ptal

kargakis commented on pkg/kubectl/cmd/util/factory_object_mapping.go:
```diff
@@ -271,6 +268,16 @@ func (f *ring1Factory) LogsForObject(object, options runtime.Object) (*restclien
 	}
+
+	sortBy := func(pods []*v1.Pod) sort.Interface { return controller.ByLogging(pods) }
+	pod, numPods, err := GetFirstPod(clientset.Core(), namespace, selector, 20*time.Second, sortBy)
```
We really need to make the timeout configurable. Spawn an issue?

soltysh replied:
The same applies for `AttachablePodForObject`, I've created https://github.com/kubernetes/kubernetes/issues/41786 to track it.

deads2k commented:
/lgtm
/approve

deads2k commented:
It's a minor refactor and the right choice for federation too.

madhusudancs commented:
/approve

soltysh commented:
This is approved and lgtm, I'll apply the label manually.

lukaszo commented:
Is there any reason to not include daemon sets here?

fejta commented:
@k8s-bot kubemark e2e test this https://github.com/kubernetes/test-infra/issues/2012

soltysh commented:
> Is there any reason to not include daemon sets here?

@lukaszo I remember having some doubts about Daemon Sets and Cron Jobs so I didn't added right now. And as long as CJ I'm quite positive it's hard to justify how to do it properly you're free to reason about Daemon Sets.

soltysh commented:
@k8s-bot kops aws e2e test this

soltysh commented:
Rebased.

soltysh commented:
Re-applying lgtm label, since this was just mechanical rebase.

soltysh commented:
@k8s-bot unit test this

soltysh commented:
@k8s-bot gce etcd3 e2e test this

crassirostris commented:
/cc @crassirostris @piosz @fgrzadkowski

fgrzadkowski commented:
@soltysh What's the motivation for this PR? How users should use it? AFAIU the semantic is that we will show logs from a random pod within Deployment. How is it useful? To be hones I find it pretty confusing.

soltysh commented:
@fgrzadkowski this is more of a convenience when you have a simple deployment/job/stateful set. We've had similar logic in place for RC/RS already, this was only adding the same for other objects.

fgrzadkowski commented:
I see. Though I find it very confusing. So we you saying that this is intended for Deployments/RS/RC with only one instance where it's clear what this mean? Do you know what was the motivation for label selectors?

soltysh commented:
Not sure I understand which label selectors you're talking about, here. The ones used in this impl are only to get the pods owned by the controller and then pick the first pod.

kargakis commented:
[The way](https://github.com/kubernetes/kubernetes/blob/9a0f5ccb331b6bb8dc3d6e2c91d05bd2af15c1c0/pkg/kubectl/cmd/util/factory_object_mapping.go#L291) pods are picked up [is not random](https://github.com/kubernetes/kubernetes/blob/9a0f5ccb331b6bb8dc3d6e2c91d05bd2af15c1c0/pkg/controller/controller_utils.go#L688) but we may want to document it.

fgrzadkowski commented:
To be honest I find that logic even more confusing.

I was referring to the following option in `kubectl logs`:
```
  -l, --selector='': Selector (label query) to filter on.
```

Maybe it'd useful to at least let the user choose which pod they are interested in?

soltysh commented:
See https://github.com/kubernetes/kubernetes/pull/44282 for the relevant discussion.
